feat: run pq portion of ivf_pq in parallel

westonpace opened the pull request:
The IVF search launches its sub-vector search in parallel:

```
let batches = stream::iter(part_ids)
    .map(|part_id| async move {
        self.search_in_partition(part_id as usize, query, pre_filter)
            .await
    })
    .buffer_unordered(num_cpus::get())
    .try_collect::<Vec<_>>()
    .await?;
```

However, if the subvector is PQ (which it always is today) then the only part that is actually run in parallel is the load of the deletion file.  This is because an async function, without any awaits, is not actually asynchronous at all.  Or, phrased another way, don't create asynchronous functions that run a long time without an await.

This PR demonstrates a simple fix to this problem using rayon.  An alternative would be for `spawn_cpu` to use tokio's `spawn_blocking` but tokio themselves suggest this is not a preferred solution for CPU intensive tasks.  Another alternative would be to use a dedicated tokio runtime (as suggested in [Andrew Lamb (datafusion)'s blog](https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/)).  The bigger challenge her is to figure out a good balance of threads to dedicate for intensive CPU work and threads to dedicate for control plane / IO work.

However, this fix is pretty effective at the moment.  The vector_index benchmark shows a 3-4x speedup.  My profiling with end-to-end ANN search shows a 2-3x speedup (assuming a simple "pure ann" with no refinement, combined knn, etc.)

One downside of this approach is that we end up cloning the PQ & query.  This isn't particularly expensive but both of these clones can probably be avoided since the clone already happens at the IVF layer.  Since it would require mucking around with the vector index traits I've left it alone for now.

chebbyChefNEQ commented on rust/lance/src/utils/tokio.rs:
```diff
@@ -0,0 +1,19 @@
+    func: F,
+) -> impl Future<Output = Result<R>> {
+    let (send, recv) = tokio::sync::oneshot::channel();
+    rayon::spawn(|| {
```
does rayon limit parallelism in anyway? Could we add an envvar to limit the parallelism?

westonpace replied:
It defaults to one thread per CPU.  You can use RAYON_NUM_THREADS to adjust.

eddyxu commented:
Can we benchmark how does rayon behaves with multiple parallel requests, i.e., 100 queries in parallel?  Using rayon w/ tokio means  tokio scheduler is unaware of the cpu usage anymore, iiuc.

westonpace commented:
> Using rayon w/ tokio means tokio scheduler is unaware of the cpu usage anymore, iiuc.

But, to answer this specific question, yes, this means there may be some amount of overscheduling (e.g. you may at times end up with more than # cpu active threads).  With current configuration you won't have more than 2*num_cpu threads though.

> Can we benchmark how does rayon behaves with multiple parallel requests, i.e., 100 queries in parallel?

In this case you'd have one thread per core doing CPU-intensive work.  You'd also have up to one thread per core doing I/O work.  It's difficult to determine what the optimal balance should be.  With 100 queries there is probably enough sync-io work to fully occupy one or more cores and so you'd have over-subscription in this case.

wjones127 commented:
> An alternative would be for spawn_cpu to use tokio's spawn_blocking but tokio themselves suggest this is not a preferred solution for CPU intensive tasks.

I think `spawn_blocking` should be fine for now, and I'd favor that over introducing a new dependency that handles scheduling.

It does seem like a good idea to experiment withe separating the CPU and IO threadpools. And if there's a good case for using rayon for the CPU threadpool, that sounds fine with me. Though if it's not much better than tokio, might as well have just one dependency.

westonpace commented:
@wjones127 I've updated it to use tokio for the second thread pool and removed the rayon dependency.  I benchmarked with rayon, tokio, and just calling spawn_blocking directly.  All three approaches had the same performance.

I did notice that the default max number of blocking threads for a tokio runtime is very high (512) as they were originally intended for I/O tasks that have no async API (e.g. DNS resolution, writing to stdout).  So I think have a dedicated tokio runtime for CPU tasks would perform slightly better if we had a large number of concurrent queries.

Either way, by having the `spawn_cpu` helper method we can always change our approach in the future (e.g. to just directly use `spawn_blocking` on the default runtime)

wjones127 reviewed (approved):
Great. Thanks for starting this.
